### ISSUE TITLE
Better handling of authenitcation across multiple instances

### ISSUE THREAD

**Issue by jamesharris-garmin** [SanderRonde/VSCode-Gerrit#70]
I very often work with AOSP style projects that are composed of repositories across multiple gerrit top level instances that can co-exist in the same workspace.

It seems like I can only really authenticate to a single gerrit instance per workspace under the current forumation of the tool.

Would it be posisble to re-define the tool to better tie into the git API and possibly look for URL authentication via the `~/.netrc` authentication or other authentication mechanisms external to vscode settings?

**Comment by SanderRonde** [comment]
Authentication is currently not the biggest hurdle to multi-instance workspaces imo, the thing I struggle most with is how things would work UI-wise etc. I'm not entirely opposed to the idea (although it does incur quite some complexity) but there are some things where I'm not really sure how things would work. Maybe you (as a multi-repo workspace user) can tell me what you'd expect behavior to be in these situations and maybe it'll turn out to be not super hard to do:
- When you open the changes tab, what would you expect to see? All changes in all instances listed together, a sort of picker where you can pick one, only the ones for the "current" repo, determined by which file is open?
- Similarly what would commands that operate without context do, for example the change picker (in the bottom bar). Would it again show all changes or only the ones for the current repo?


So I think the main question is how would you determine the current repo, or maybe we want all changes from all repos combined together (which would probably make quite some API calls quite a bit slower). If the answer is that we do just take the current file, is it going to be a situation where you are likely to switch repos often? Do you generally stick to a repo or do you hop back and forth?

**Comment by Mattadore** [comment]
As another multi-instance user, I'd say just having `gerrit.gitRepo` just be a list of paths to care about would be awesome. I don't personally deal with the multi-auth issue, but supporting multiple gerrit instances under a single workspace would be incredibly helpful.

> Do you generally stick to a repo or do you hop back and forth?
I hop around constantly, and generally just expect things to operate on the repo of my currently selected file (all of my custom actions behave like this). Though, if it only cares about explicitly specified ones it's not so bad, I only really work on 3 or 4 repos within the workspace regularly. I expect that people using this for multi-gerrit in one workspace are probably working on something huge like AOSP.

> Would it again show all changes or only the ones for the current repo?
Again, my strong preference is just whatever the repo of the currently selected file is, that's how all my other workflows behave.

> maybe we want all changes from all repos combined together
Please no, pretty please

**Comment by felipecrs** [comment]
If possible, I would recommend to use VS Code's Authentication API instead:

https://www.eliostruyf.com/create-authentication-provider-visual-studio-code/

**Comment by SanderRonde** [comment]
Alright these are some good answers to my questions :)

I think this does sound like a valid thing to support. To make it work I would like to have such a repository to test it out on. Since I'm not too knowledgeable on this I'll ask you guys. Which (open source) repo is a good example of what you'd like to have supported by this extension? And what does the setup look like? Do you just have a single root folder that you clone all the gerrit instances into? Do you use VSCode's multi-workspace feature? I did briefly look into the AOSP project but it looks like that takes up a massive amount of storage so that's probably not ideal.

**Comment by jamesharris-garmin** [comment]
I don't know if there is a specific open source project that would cover this use case, but it might be possible to fake it simply by cloning a subproject of AOSP and then getting another simple sub-project like git-review from the openstack guys (opendev.org). These could be in either a multi-root workspace or a single folder with multiple projects (thought the latter case is probably more likely).

It would simply be a case of being able to list witch gerrit servers can log into and then matching the project remote servers against that list of domain names. Then you present a tree view of which projects are associated with which gerrit instance we have configured.

**Comment by Mattadore** [comment]
I mean, a simple repro is just having a folder with 1-2 multi-level nested subfolders that are Gerrit repos, I don't know if it's much more complex than that. eg: some root workspace, then a/b/somerepo + a/b/someotherrepo + c/d/somethirdrepo. I assume if you get that working it'll probably scale fine, and I can test it on AOSP.

> Do you use VSCode's multi-workspace feature?

Different people handle this differently, unfortunately.. there's not a great standard around it. From what I can tell most people do, but I don't, I just have one giant workspace for everything. I assume multi-workspace would just configure this as multiple individual instances, one on each workspace, which I guess would have solved my problem? But I don't really like multi-workspace workflow very much, it's clunky. 

>  I did briefly look into the AOSP project but it looks like that takes up a massive amount of storage so that's probably not ideal.

...yeah 😅

**Comment by SanderRonde** [comment]
Alright it's taken a while (and some 4.5k [changed lines of code](https://github.com/SanderRonde/VSCode-Gerrit/commit/c916d46e6ef6786d7c9b50352a13c38833744ad7)) but I think I've got it. I've attached a build of what I've got now. It seems to work pretty well from what I can test. See the linked commit's commit message for some more details.

To install the file: rename from a `.zip` to `.vsix` (github doesn't allow uploading `.vsix` files), then go to your editor, to the `extensions` tab, click the dots on the top-right, click "install from vsix" and pick it there.

[vscode--gerrit-1.2.41.zip](https://github.com/user-attachments/files/16432380/vscode--gerrit-1.2.41.zip)

**Comment by Mattadore** [comment]
Nice :smile: really appreciate the support!!!

I will check it out this week and see how it goes.

**Comment by jamesharris-garmin** [comment]
In trying to configure this I immediately see a problem where I am trying to clone from an ssh:// url and need to specify an alternate http url. (I.e.) the root for the code review URL is `https://<server-name>/code-review/` but the repository is cloned from `ssh://<server-name>`

Should I put the https url in there?

**Comment by SanderRonde** [comment]
Yes. The HTTP URL is the URL that will be used by the [HTTP API](https://gerrit-review.googlesource.com/Documentation/rest-api.html). I'm not sure how things are configured in your case, the default setup has both be the same, but I'm pretty sure the URL you use to view your dashboard and such is the one that'll host the API too.

**Comment by jamesharris-garmin** [comment]
the ssh server is just the domain name .but the gerrit application is hosted form the 'code-review' path.

so the sign in workflow would need to be able to say: oh you have an ssh server, set the root of the gerrit application (hit enter if it is the same)

**Comment by jamesharris-garmin** [comment]
also if you could have an example layout of how to manually do this in the settings that might be helpful

**Comment by SanderRonde** [comment]
Ah because it prefills it with the host in the `.gitreview` file in the "setup credentials" command. Yeah I agree, I'll add that later. Indeed a bit confusing to prefill with possibly the wrong value.

**Comment by SanderRonde** [comment]
An example:

```
"gerrit.remotes": {
	"https://<server-name>/code-review/": {
		"username": "...",
		"password": "..."
	}
}
```

**Comment by SanderRonde** [comment]
Ah I just realized that I accidentally threw out support for having a different URL from the SSH URL in the current multi-repo approach. Will bring it back again and let you know once I did

**Comment by SanderRonde** [comment]
Ah turns out it was stil in there but just in another way. I've re-built the extension with better support for it. Now it is a part of the "Enter credentials" command, as well as being part of the `gerrit.remotes` setting.

[vscode--gerrit-1.2.41.zip](https://github.com/user-attachments/files/16502991/vscode--gerrit-1.2.41.zip)

**Comment by Mattadore** [comment]
`Activating extension 'SanderRonde.vscode--gerrit' failed: Cannot read properties of undefined (reading 'host').` :(

**Comment by SanderRonde** [comment]
Hmm I can't seem to replicate the error. What are your settings?

**Comment by Mattadore** [comment]
Is the code assuming that we have gitconfig files? It seems to make that assumption, and I have none. Could that be the issue?

**Comment by Mattadore** [comment]
```
    "gerrit.remotes": {
        "default": {
            "url": "https://android review url",
            "username": "someusername",
            "password": "somepass",
        },
    },
```

**Comment by SanderRonde** [comment]
Ah yeah it is indeed somewhat making the assumption that you have a `.gitreview` file (and I can reproduce the issue when I remove it myself). Does gerrit still work for you without such a file? And do you know if this is a common setup?

**Comment by Mattadore** [comment]
Yeah, I have never had a file and everything worked fine before the latest update. The credentials are managed in a different system. I believe this is a fairly common setup, in my experience.

**Comment by SanderRonde** [comment]
Ah ok. I've changed the setup to be able to catch not having a gitreview file. I think it should work for you then.

[vscode--gerrit-1.2.41.zip](https://github.com/user-attachments/files/16531923/vscode--gerrit-1.2.41.zip)

**Comment by SanderRonde** [comment]
Yeah fair point, will try to do that some time soon. Indeed shouldn't be terribly hard.

**Comment by Mattadore** [comment]
```[SanderRonde.vscode--gerrit]Maximum call stack size exceeded
$onExtensionRuntimeError @ mainThreadExtensionService.ts:78```

<img width="334" alt="image" src="https://github.com/user-attachments/assets/c860109b-ddc5-4fec-ac0a-f17b1487b825">
It seems to be stuck in an infinite loop

**Comment by SanderRonde** [comment]
Oops my bad, indeed looks related to that. Here's another build.

[vscode--gerrit-1.2.41 (2).zip](https://github.com/user-attachments/files/16564402/vscode--gerrit-1.2.41.2.zip)

**Comment by SanderRonde** [comment]
For those interested, I've completely removed the reliance on the `git-review` tool and, where possible, the `.gitreview` file. Here's a build of the extension with those changes. (CC @felipecrs)

[vscode--gerrit-1.2.41-2.zip](https://github.com/user-attachments/files/16570086/vscode--gerrit-1.2.41-2.zip)

**Comment by felipecrs** [comment]
That's really nice!

**Comment by Mattadore** [comment]
I will check it out soon, I have been a bit busy, but seems very promising :D

**Comment by SanderRonde** [comment]
I've not released either of these changes yet. I'm kind of hoping you guys can test out the multi-repo variant since I can't test it in my day-to-day workflow. After that one has gone live the git-review change will probably also go live soon after since I was able to test that myself.

**Comment by felipecrs** [comment]
Got it. I don't personally use multi-repo feature, I think.

But I will try out the vsix and let you know. I'm specially interested on the git-review changes.

**Comment by Mattadore** [comment]
```
Tried to run "git config --get remote.origin.url", but failed
Stdout: 
Stderr: 
Error: Command failed: git config --get remote.origin.url
```

soooooooooo I don't use "origin"
like, I could add "origin" but it would need to be done manually for each repo.
I don't think it is unreasonable to assume "gerrit" remote exists, but better would probably be to just have a setting for default remote name. Ideally once I get this working,  I could export my config to other Android users and have it "just work" ™️

**Comment by felipecrs** [comment]
If none is set, perhaps trying first "gerrit" then "origin" would be a good default?

**Comment by felipecrs** [comment]
(I don't use "gerrit" in my git remotes, I only use "origin").

**Comment by Mattadore** [comment]
FWIW almost none of my repos use "gerrit" or "origin" and that would immediately break, I actually retract my prior statement about being able to assume "gerrit" exists. Also, it is trying to access data from the remote itself which immediately fails because we don't use "ssh:" or "https:" but a weird non-standard prefix. The thing is, all of this worked before the multi-repo split, and still works on the current public version, so I am wondering what I can do to just make it behave like it did before.

**Comment by Mattadore** [comment]
Actually, it seems like this is maybe just a.. bug? It is trying to figure out what the gerrit server is in order to access "config/server/version" , but it's moot because I already passed the server URL to the config so it should just use that, I think it doesn't need to know what my remote is or any of that.

**Comment by Mattadore** [comment]
It's no rush, I appreciate all the support :)))

**Comment by SanderRonde** [comment]
Some changes to the git-review related changes in this new version. I've made it go down a list of remotes, falling back to the first remote in `git remote` when there is no match.

You should be able to set the correct API URL by setting the `url` field of the new `gerrit.remotes` field. You'll probably want to use `default` as a key for the setting (at least while testing).

[vscode--gerrit-1.2.41.zip](https://github.com/user-attachments/files/16733559/vscode--gerrit-1.2.41.zip)

**Comment by Mattadore** [comment]
It is still using the git remote for review, not the "url" param.

I am actually not sure where the "url" config param is used at all in the code, but am maybe just blind?

**Comment by SanderRonde** [comment]
Yes I think that makes sense or am I missing something? The git remote is the one that git-review changes are git-pushed to. The URL is used for the HTTP API, which is where the extension gets most of its data from.

**Comment by Mattadore** [comment]
It appears to be trying to make https requests to "https://" + remote + repo "/config/server/version" which will not work, and it was not doing previously. I cannot share the exact string, but it's not a valid http request and the remote does not even have a TLD, we use a weird auth provider

**Comment by SanderRonde** [comment]
I did change quite a bit about the structure of the API in order to make multi-instance usage work. So do expect some things to be broken and to not work, mostly because it's quite hard for me to test this against a real-life use case. 

I think I did fix the bug you're mentioning in the package below, I'm not 100% sure though (since I can't really verify). Could you check for me whether this helps? If not, a few things you can do:
- Use the "Setup credentials" command to set it up from scratch, that might help
- Check the devtools and its logs. There'll be a bunch of `Dev: ...` logs there. It would be a great help if you could share those (of course feel free to censor them)

[vscode--gerrit-1.2.41-3.zip](https://github.com/user-attachments/files/17002236/vscode--gerrit-1.2.41-3.zip)

**Comment by Mattadore** [comment]
It works! Kind of. It successfully loads all the ongoing reviews under a given repo, and if I change repos it then loads those. However, comments don't work at all and I have no idea why, the devtools aren't returning any issues nor is the extension output. Importantly it's not just per-file comments, even opening an ongoing review does not load comments, they just don't seem to work at all.

It is also bit slow, though I think there are some possible fixes, as it seems to be an issue of just "a ton of requests at once". I think batching requests into one mega-request might be possible, as the webui seems to do this, and we can also probably just lower the amount of changes it fetches per-category at once if that doesn't work.

**Comment by Mattadore** [comment]
I did not mess with the auth config at all, the one I was trying previously immediately worked on the new version.

**Comment by SanderRonde** [comment]
I didn't get to the comments issue yet but I did manage to speed up the requests a lot. While it wasn't a matter of syncing up requests (there's only one request that supports multiple changes), it was a matter of removing unnecessary requests by rewriting some code. Loading the changes panel and expanding changes is now much faster :)

Thanks for the suggestion btw, I also released this optimization to the main extension branch

[vscode--gerrit-1.2.41-4.zip](https://github.com/user-attachments/files/17175471/vscode--gerrit-1.2.41-4.zip)

**Comment by SanderRonde** [comment]
I tried to replicate the no-comments issues but they do work for me. Some more questions:
- Are you able to place any comments? What happens to them?
- Do you see comment icons next to files in the changes panel on the left?
- Are you sure they're not resolved and that's why they're initially collapsed?

After these questions the best thing to do is probably for me to make a debug build for you that logs to the console which comments are loaded from the API and to figure out why they're not displayed.

**Comment by Mattadore** [comment]
I got it to load once, by like.. re-loading it a couple of times and going through the change menu directly. The comment icons do not normally show up, and the gutter to add comments isn't there unless I go through the change menu directly, but it seems to be _able_ to load comments, it just isn't.

I am starting to suspect this is related to call volume, as the client seem to be sending a lot of duplicate calls and the server is responding by rate limiting them and failing some of the transactions. I think the calls to get comments might just be getting dropped by the server rate limiting?

**Comment by Mattadore** [comment]
It seems the one time it worked may have been a fluke and I can't repro???? 

> Are you able to place any comments? What happens to them?
No

> Do you see comment icons next to files in the changes panel on the left?
No

> Are you sure they're not resolved and that's why they're initially collapsed?
Yes, no comments resolved or unresolved.

The one really confusing thing is "Patch-level comments" consistently work, reproducibly, every time.. though they only show up if I manually select the "Patch-level comments" tab.

**Comment by SanderRonde** [comment]
I've found the issue and fixed it. You know the drill, I've attached the fixed version again.

[vscode--gerrit-1.2.41-5.zip](https://github.com/user-attachments/files/17378882/vscode--gerrit-1.2.41-5.zip)

**Comment by Mattadore** [comment]
Oh, I was actually coming back to say it magically started working but I will check this out

**Comment by Mattadore** [comment]
fwiw I have not forgotten about this, I was just in busy season at work and intend to come back to this ~~~soonish

**Comment by jamesharris-garmin** [comment]
I don't think that is necessarily the case, it should use whatever auth git uses. but that depends on how some companies may be internally mirroring the AOSP project. which I admit is harder to test for.  I wonder if it would be easier/better to piggyback on whatever configuration the workspace host system is configured? Not sure how this would effect remote workspaces though.
